### ISSUE TITLE
Clear problems using a task

### ISSUE THREAD

**Issue by tksuoran** [microsoft/vscode#50448]
Is there any other way to clear problems window, other than using "Reload Window" command?
I would like to perform this from a task. My tasks generate new entries to problems, and I would like to clear the old problems first.

**Comment by dbaeumer** [comment]
There is not need to do this. VS Code does it automatically. Do you have steps where this doesn't happen. Please note that VS Code only does this for problems captured by a problem matcher for each task. So if you have two tasks with two different problem matchers running them only clears the onces belonging to the matcher.

**Comment by tksuoran** [comment]
I have several dozen tasks and together their problem matchers have thousands of hits. I do need to clear problems between tasks in order to be able to examine problems of a specific task.

Alternatively it would okay if task/problem matcher could associate a "code" and problems would group issues with this code and I could fold/unfold groups individually.

**Comment by Tallyb** [comment]
Some examples of non clearing problems: 
 - problems pointing to a deleted file.  
- Problems pointing to a gitlens revision file

**Comment by dbaeumer** [comment]
@Tallyb both of these kind of problems should be removed by the one creating them in the first place either when the task is rerun or in a case of watching when the file events are sent out.

**Comment by aaroncraig10e** [comment]
This is a prime example of a developer deciding for the community what best practice is, and it's horribly arrogant and frustrating.

VSCode should not dictate any given workflow.  It should support my workflow, because as an intelligent engineer, I know the best workflow for my use case.

Refusing to add what must be a trivial feature simply because you as the maintainer of the project wouldn't use such a feature smacks of an elitism that has no place in this community.

This attitude ruins an otherwise amazing package, and frankly leaves me wanting to look elsewhere for an IDE merely on principle.

The fact is, there are many legitimate use cases for ignoring errors at any given time, all of which have been listed either here or on one of the other issues raising this problem:
- I've closed those files as they do not impact my current task
- They are errors that I cannot or do not want to fix in the present moment
- They are not errors, because the offending file has been deleted

In each of these cases, the stale error output in the problems pane creates noise that makes it hard to find my actual issues, creating a frustrating user experience.

Please add this simple feature!

**Comment by dbaeumer** [comment]
Closing the issue was not an active developer decision. The issue got closed because we have ask for more information / reproducible steps to better understand what is happening and if that information doesn’t arrive in 14 days then our bot closes the issue automatically (for more information about how we do issue management pls [see]( https://github.com/Microsoft/vscode/wiki/Issue-Tracking#ongoing-issue-management))

In this particular case some information was provided and we should have dug deeper to better understand what is frustrating by asking for a better example of what tasks are executed to find a solution to solve @tksuoran´s problem.

In general if something like this happens it is enough to send a ping by adding a comment in the issue and ask what information we still expect.

**Comment by tksuoran** [comment]
I have couple of dozen tasks. These are reports from different static analysis rule checkers. Together they report thousands of issues spread to a large codebase. Some of the reports contains hundreds of issues, some report only few or just one.

I want to examine these reports one rule at time. So when I execute a task to view such report, I want to throw away all other reports. So in my case I would like to use wild card or some other way to simply throw away _all_ previously shown issues.

**Comment by aaroncraig10e** [comment]
Just ran across another significant scenario -- switching git branches.  In this case, I end up with errors being reported for inexistent files.

**Comment by dbaeumer** [comment]
@aaroncraig10e this clearly sounds like a bug for me. Can you please open a separate issue so that we can track and fix it.

**Comment by dbaeumer** [comment]
@sandy081 do you have an issue for [this](https://github.com/Microsoft/vscode/issues/50448#issuecomment-424269795) for the problems pane.

This issue is for clearing problems using a task run which in @Gruntfuggly use case will not help.

**Comment by Gruntfuggly** [comment]
The reason I added my case here is because the other threads have been closed or locked.

**Comment by sandy081** [comment]
@dbaeumer No. Not sure if it is intended but open for discussion. May be in this case Markers Service can clear the problems automatically if the extension goes down.

**Comment by aaroncraig10e** [comment]
FWIW I'm only seeing persisting errors (when files close) for `[ts]` errors.

**Comment by dbaeumer** [comment]
@sandy081 the case of an extension crashing should definitely be handled by use and we should clear the markers its owns. Same if the extension host goes down. I will open a issue for it.

**Comment by tksuoran** [comment]
I have a task with a problem matcher which parses task output. If there is any way to specify how/when the problems (found by the problem matcher in the task) would be cleared, please let me know. Or just implement the command which clears all problems generated by custom problem matchers.

**Comment by dbaeumer** [comment]
@tksuoran currently the problems are cleared when the task is rerun.

**Comment by creinarz** [comment]
I ran into this problem while compiling a program on a remote system. The task executes make on the remote system and matches the output with a problem matcher. Errors and warning from the compiler are never cleared. Here's my task.

```
        {
            "label": "build (aria)",
            "type": "shell",
            "command": "ssh",
            "options": {},
            "args": [
                "creinarz@xxx.xxx.xxx.xxx",
                "cd /home/creinarz/VSProjects/myproject/build &&",
                "make -j8 -w 2>&1 | mappath home:Users"
            ],
            "problemMatcher": {
                "fileLocation": [
                    "absolute"
                ],
                "pattern": {
                    "regexp": "^(.*):(\\d+):(\\d+):\\s+(warning|error):\\s+(.*)$",
                    "file": 1,
                    "line": 2,
                    "column": 3,
                    "severity": 4,
                    "message": 5,
                }
            },
            "dependsOn": "sync (aria)",
            "group": {
                "kind": "build",
                "isDefault": true
            }
        }

```

**Comment by adam-ah** [comment]
Seems simple enough to add, I agree.
I have the same issue with tslint errors lingering after delete / close.
I agree, _ideally_ plugins would remove these errors on close/delete/whatever, but it's actually much simpler to clear the problem window then try to fix all plugins that do not re-report the problems, including custom tools that people hacked up for themselves.
In fact, all IDEs I know do support this. Why the resistance?

**Comment by dbaeumer** [comment]
@nawfalhasan can you please provide me with steps to reproduce this. When a task reruns it tries hard to clear old problems it generated. 

@adam-ah this item is about clearing problems a task created. I can't provide a task that clears problems it doesn't own or are at least generated by another task. Major reason is that problems have an owner.

If you want to have an action in the problems panel that allows you to clear all problems please file a separate issue since this is not related to tasks.

**Comment by Gruntfuggly** [comment]
@dbaumer Lots of issues have been created asking for something to clear the problems, but they just get closed with "it's the extension's problem" response. This is the only one I can find that's related and still open...

Please could you explain why there is so much resistance to adding such a simple feature?

**Comment by adam-ah** [comment]
@dbaeumer I do not agree. The very first line of the original post:

> Is there any other way to clear problems window, other than using "Reload Window" command?

This bug is indeed about clearing the problem window. Why are we facing resistance against such a basic feature?

Also, I agree with @Gruntfuggly, many tried to create other issues to ask for a Clear Problems comman, but all such requests consistently get closed down (eg., #15524, #29169, #66343) mainly blaming the plugin that does not correctly clear them; this is the only one that survived and I'm really hoping @sandy081 doesn't close this one as well, making it impossible for the community to discuss it.

I think it is more realistic request to add a "Clear Problem Window" command to VSCode rather than waiting for all current and future extensions to work perfectly. All other IDEs and all other windows (Output, Debug console, Terminal) support this, why fight adding this to Problems? I get it, you don't need it or even like the idea, but you aren't the main _user_ of the VSCode either.

**Comment by dbaeumer** [comment]
@adam-ah but the second line clarifies:

> I would like to perform this from a task. My tasks generate new entries to problems, and I would like to clear the old problems first.

and I accept with the comment here https://github.com/Microsoft/vscode/tree/HEAD@%7B2018-07-05T08:26:38Z%7D that this is something tasks can do.

The reason why tasks can do this is that the task system creates these problems and can delete them.

In VS Code problem model problems are owned by the one creating them. An extension could therefore always only apply the delta (e.g.remove one and add another one) instead of always computing all errors and pushing them into the model (the task system by the way does some delta optimizations as well). So if a user clears all problems he might miss out on new problems even if he changes files since the extension computing them might still think old problems are present (for example during incremental builds).

However I have to say that thinks got improved in the API since discussions started about clearing problems and extensions can read diagnostics. But there is still no event that signals that all problems got cleared which IMO is a prerequisite to provide such a feature.

But all this is not a task problem and shouldn't be addressed through this issue. So I would like to suggest that you create a new issue and we discuss the approach outlined in this comment there since it needs more than the task system can provide.

**Comment by darlanalves** [comment]
There you go.
https://github.com/Microsoft/vscode/issues/66982

As @dbaeumer stated, please discuss the matter in the new issue if it concerns adding a button to clear the Problems pane.

**Comment by bobbrow** [comment]
> @aaroncraig10e no, since TS is following the recommendation of VS Code. Compiler / Linters that have a project concept and inter file dependencies are asked to keep problems around even if the file is closed. The reason is that a problem in file A.ts can be fixed by changes in file B.ts. If the problem would be removed if A.ts closes the user had no indication that his change in file B.ts actually fixed the problem.

If I close a TS file, I want the [ts] problems to go away.  If I'm trying to fix a problem in A.ts by editing B.ts, I'm going to keep both files open. 

@dbaeumer, when was this recommendation made?  I disagree with this direction.  I work on the C/C++ extension and we used to get complaints from our customers about leaving orphaned problems around when files were closed.  Why would we want to go back to that behavior?  It will just mean more support tickets for us.

**Comment by dbaeumer** [comment]
@bobbrow it was always that way although TS itself has problem following it. How would you otherwise detect something like this:

- user opens file A edits it and it creates problems in file B
- file B is not open hence no errors shown to the user

I as a user would expect that a system tells me about the problem in file B.

Can you explain when you leave orphaned problems around. Is it because VS Code has not enough API so you can clean them up or is it because it is hard to detect for you that a problem is actually fixed.

**Comment by bobbrow** [comment]
I actually don't want the TS extension to show problems for B.ts if I never opened it.  I will wait for the compiler to tell me something is wrong.  Likewise for C++, it would be too expensive for us to lint every unopened file in the workspace if a user made a change to A.h which negatively affected some B.cpp.  I let the compiler tell me when I break something.  This is the equivalent experience I am used to in Visual Studio.

We only had orphaned problems when there was a bug in our extension.  We fixed those issues and our customers are happy now. This issue has only come up because one of our customers now wants us to leave squiggles around after closing files which is when I noticed the functionality change with TS.

I'm fine with exposing a setting in the C++ extension that lets us control whether problems persist after closing files, but I think the setting would more appropriately live in VS Code.  Either way, based on the feedback we've received and my own personal opinion, I would not recommend this behavior be the default.

**Comment by dbaeumer** [comment]
I actually do want a tool to show me errors as soon as they come into existence not only after a build :-).

**Comment by bobbrow** [comment]
Unfortunately, that is impossibly expensive for the majority of C++ projects.  In addition it would be overly noisy while I am in the middle of editing code (e.g. the linting kicks in while I have a mismatched curly brace or am refactoring a class and suddenly 1000+ problems show up across all files in the project).

In my opinion it makes more sense to let the user decide when is the appropriate time to rescan the entire project (e.g. when I'm ready to compile it).

**Comment by dbaeumer** [comment]
> Unfortunately, that is impossibly expensive for the majority of C++ projects.

I can see that this is challenging for C++

> In addition it would be overly noisy while I am in the middle of editing code (e.g. the linting kicks in while I have a mismatched curly brace or am refactoring a class and suddenly 1000+ problems show up across all files in the project).

Good compilers handle these cases nicely by only checking or showing sematic errors if there are no syntax errors.

**Comment by bobbrow** [comment]
> Good compilers handle these cases nicely by only checking or showing semantic errors if there are no syntax errors.

Sure, that handles the mismatched curly brace case, but not the case where I'm cutting and pasting variables/full functions/classes, reordering arguments, or other syntactically correct but semantically incorrect refactors.

Not even TypeScript turns off semantic errors for syntactic errors though, so it would seem that this is more challenging to implement than it sounds.

![image](https://user-images.githubusercontent.com/12818240/53654849-94c4ba80-3c03-11e9-81d4-18bf37e55163.png)

**Comment by developer1011** [comment]
Currently the only workaround is to use the 'reload' extension  
https://github.com/natqe/reload

**Comment by felix9** [comment]
Repro steps:
I have a typescript project with one tslint error.
I do Terminal->Run Task.. "npm lint", which runs "tslint" (v5 output)
The Problems pane shows the one lint error.
I fix the lint error and save the file.
I repeat the Run Task.
The terminal shows "All files pass linting".
The Problems pane still shows the lint error.

I expect the Problems pane to clear the messages from the Run Task when I rerun the task, but it seems to clear the messages only after it parses the results.

**Comment by dbaeumer** [comment]
@felix9 do you have a repository that I can clone that demos this ? And which tslint extension are you using?

**Comment by resk8** [comment]
I still cannot believe there isn't a way to clear the problems tab when re-running a task YET. How hard is to add a new task property such as "ClearProblemsAtLaunch":"true" ?

**Comment by stewhi** [comment]
Has there been any movement on this? I'm encountering the following situation:
Run a watch task with a problem matcher
Introduce a bug
Problem matcher puts an error message on the problems pane
Fix the bug
Problem matcher runs, no errors in the terminal
Problems pane still has the error reported, and the "red squiggle" is still on screen where the bug used to be.
Restarting the watcher clears out the problems pane and red squiggle, but that's less than ideal.

**Comment by dbaeumer** [comment]
@stewhi this issue is about to clear problems from Task A from Task B. This is not about clearing problems generally. This is tracked here: https://github.com/microsoft/vscode/issues/66982

Can you provide me with a setup were you see this. In most cases this is caused by not well configured problem matchers.

**Comment by mttindustries** [comment]
I'm in an agreement with everyone else in this thread. I've opened up several files as I work on multiple projects and it just keeps a log of all the past problems WITH EVERY SINGLE FILE. They don't disappear after closing or fixing them. I have over 1k of problems now because of every-time I forgot a semicolon and it's duplicated 10 fold with all the back-history of each page. Why isn't there a feature to say right click on and mark as solved or clear all? It's the one thing that's letting down this program and is making me contemplate switching back to dreamweaver

**Comment by mttindustries** [comment]
> Currently the only workaround is to use the 'reload' extension
> https://github.com/natqe/reload

I tried that, it didn't work for me

**Comment by Gruntfuggly** [comment]
#80419 is a great example of why this issue is so frustrating. **The file creates a problem which is not removed when the file is closed.** The argument that it is always that it's up to the extension providers to clear up their problems is a bit moot when the main application itself can't even do it,

Please please please can we just have a button to clear them!

**Comment by alexr00** [comment]
How would folks who are passionate about this feel about having a command that clears task related problems? I don't want to make a task that does this because we don't have the concept of a universally contributed task. i.e. tasks are contributed by an extension or written by an individual users, not provided by core VS Code. https://github.com/microsoft/vscode/issues/77214

**Comment by adam-ah** [comment]
I can't believe this was closed down again after many previous attempts to convince MS to fix this (eg., #15524, #29169, #66343).
What the heck is going on?

**Comment by alexr00** [comment]
I do want to implement this functionality, I just think that it would be better as a command, not a task. Please see the comment I left: https://github.com/microsoft/vscode/issues/50448#issuecomment-539414911

**Comment by jeiea** [comment]
I've set 'owner' of task, and it seems to clear same task's problems. I don't know why but satisfied with this for now.

**Comment by mttindustries** [comment]
> I've set `owner` of task to something other than `external`, and it seems to clear same task's problems. I don't know why but satisfied with this for now.

How do you do that?

**Comment by jeiea** [comment]
```jsonc
{
  "version": "2.0.0",
  "tasks": [
    {
      "label": "build",
      "type": "shell",
      "command": "chcp 65001; stack build --fast",
      "group": {
        "kind": "build",
        "isDefault": true
      },
      "problemMatcher": {
        "pattern": {
          "regexp": "^(?:.*?> )?(.*?):(\\d+):(\\d+): (.*?):",
          "file": 1,
          "line": 2,
          "column": 3,
          "severity": 4
        },
        "owner": "haskell" // <- here
      }
    }
  ]
}
```
Excerpted from [here](https://gist.github.com/jeiea/6a141aa7d0509f5ec9aeb9ce8ef992d4)
